### DISCUSSION TITLE
Data source information enrichment in HTTP headers of /api/ds/query (QueryMetricsV2) requests

### DISCUSSION THREAD

**Discussion by zhichli** [grafana/grafana#55780] [Data Sources]
> **Latest edit: this PR #56832 invalidate the assumption that we can add plugin type header in frontend for `api/ds/query` path since the #56832 PR allows different data source types in single request. But I describe a more general ask in the [comments below ](https://github.com/grafana/grafana/discussions/55780#discussioncomment-4065099) with [some other proposals](https://github.com/grafana/grafana/discussions/55780#discussioncomment-4065262).**

Currently, certain headers are injected into data query requests when backend service is called such as `X-Grafana-Org-Id` in https://github.com/grafana/grafana/blob/2bbc7c40e9fe261b2fe43bd92685d62ec7ed0fd8/public/app/core/services/backend_srv.ts#L180. It would be useful to have plugin/data source information also enriched in the headers to help troubleshooting and aggregate usage insights especially when running a reverse proxy in front of Grafana (common hosted scenario). The proxy can extract plugin/data source information from headers to collect business insights or helpful diagnostics info. 

As illustrated below, the `x-test-plugin-type` shows the data query request contains queries for `prometheus` data source:
![image](https://user-images.githubusercontent.com/57812115/192397370-7c58c103-8f02-4c1b-9480-6539bd2a471e.png)

The feature ask is that 
- adding a header (such as `x-plugin-type`) containing the data source type
- adding a header (such as `x-plugin-query-count`) containing the data queries count in that single request

Theoretically, data source type and queries count could be extracted from [`options: BackendSrvRequest`](https://github.com/grafana/grafana/blob/2bbc7c40e9fe261b2fe43bd92685d62ec7ed0fd8/public/app/core/services/backend_srv.ts#L171)

**Comment by marefr** [comment]
We already have existing Prometheus metrics exposed for outgoing datasource HTTP requests from Grafana, i.e. `grafana_datasource_request_total`, `datasource_request_duration_seconds`, `grafana_datasource_response_size_bytes`, `grafana_datasource_request_in_flight` together with label `datasource` containing the name of datasource in question. You can't use these?

What about datasource queries not originating from the browser, e.g. Grafana Alerting, Recorded Queries, you're not interested in collecting business insights or helpful diagnostics info for these?

What about datasource queries going through HTTP API/not originating from Grafana browser app, they will not have any of your suggested headers set.

What URL's are you imagine this to be supported for, i.e. only `/api/ds/query`? I imagine only this one based on the title, but want to confirm.

> Currently, certain headers are injected into data query requests when backend service is called such as `X-Grafana-Org-Id` in

That one is global/set for any request going through backend_srv.ts and is for the Grafana backend to understand what the current org id of the HTTP request is.

> Theoretically, data source type and queries count could be extracted from [`options: BackendSrvRequest`](https://github.com/grafana/grafana/blob/2bbc7c40e9fe261b2fe43bd92685d62ec7ed0fd8/public/app/core/services/backend_srv.ts#L171)

Theoretically yes, but compared with `X-Grafana-Org-Id` which is globally set for all HTTP requests, the `x-plugin-type` and `x-plugin-query-count` is not, they're only available in a certain context, e.g. when executing datasource queries. I think we need to handle this somewhere in between a datasource plugin and the backend_srv to be able to consider accepting this.

**Reply by zhichli** [reply]
Thanks for the insightful comments @marefr. You brings up a very good point that there could be multiple sources sending out data query requests besides frontend browser app (I should've considered this 😅).  Here is a list of interests with different urgency in different sources:

- [x] frontend (end user facing) (must have)
This is definitely the most important one or even mostly used one. And this source reflects directly the user interactive usage on data sources, so we would like to enable this feature for this source **at least** (if other sources are not easy to enable), which could give us the most urgent need for business insights.
- [ ] from headless source e.g. call api directly from curl command line (end user facing)
This source is good to have but this should not be a big part of the usage we assume.
- [ ] from alerting engine (internal to Grafana)
This source is also good to have. Although it doesn't directly reflects user interactive usage, it can give us the data source usage from the perspective of alerting which is also very useful for alerting feature evaluation.
- [ ] from recorded query (internal to Grafana)
Our take on this is similar to alerting - good to have. We haven't enabled this feature for our customers, but once it's enabled, its usage insights should be useful as well.

And I also realize data query can come from `api/datasources/proxy` (for certain data source plugins) besides `api/ds/query`, so we are interested in **both api paths** (or including any other path we are not aware of).

We are well aware of the existing instrumentation with Prometheus metrics on data source requests. But we don't have infrastructure for metrics, which would requires a quite high cost including setting up metrics collection agent and cluster with storage solution. We do heavily relies on logs and logs gives more flexibility to us for certain queries (like join tables), so we are really requesting this feature from **logs** perspective.

**Reply by zhichli** [reply]
In terms of what fields we want for logging data query requests, here is a list (must-haves are checked):

- [x] datasource type (for both usage insights and troubleshooting)
- [ ] datasource name or id or UID (for identifying DS when troubleshooting)
- [x] user email (if user context exists) (for usage insights, we want to count different queries for different users)
- [ ] query source
  - dashboard id or UID (if sent from dashboard panels) (for usage insights and troubleshooting: we want to count how many dashboards use certain data source)
  - from headless caller or not
  - from alert or not
  - from recorded query or not
- [ ] request duration (for latency insights and troubleshooting)
- [ ] status code or isSuccessful (for reliability insights and troubleshooting)

I admit I'm not perfectly familiar with the data source plugin code part yet, but after diving a bit more into the codebase, I'm thinking of 2 alternatives for meeting our need on different levels. But you probably have better proposals.

- printing logs when [meta analytics backend](https://github.com/grafana/grafana-enterprise-mirror-shared-microsoft/blob/a255ecc98ede30440010bd855839d8af2b912c16/src/pkg/extensions/analytics/summaries/summaries.go#L93) writes events
  - Pros:
    - both [`data-request`](https://github.com/grafana/grafana/blob/bdb8fe3b025f8f5eb11d85f8287901e73ac4b35b/packages/grafana-runtime/src/analytics/types.ts#L41) (which is what we are asking) and [`dashboard views`](https://github.com/grafana/grafana/blob/bdb8fe3b025f8f5eb11d85f8287901e73ac4b35b/packages/grafana-runtime/src/analytics/types.ts#L40) (which is also useful) can be covered
    - [`Event` model](https://github.com/grafana/grafana-enterprise-mirror-shared-microsoft/blob/a255ecc98ede30440010bd855839d8af2b912c16/src/pkg/extensions/analytics/eventmodels/models.go#L43) contains most of the fields we ask
  - Cons:
    - it only covers data query requests from frontend browser app
    - no status code is available since it's logged before response
    - there is a duration in [`Event` model](https://github.com/grafana/grafana-enterprise-mirror-shared-microsoft/blob/a255ecc98ede30440010bd855839d8af2b912c16/src/pkg/extensions/analytics/eventmodels/models.go#L43) but it doesn't cover e2e AFAICT
- printing logs in [`query` service](https://github.com/grafana/grafana/blob/a83fdc6b879127d506f0ffc3639a25cdf36bf246/pkg/services/query/query.go#L224) before sending out to individual data source
  - Pros:
    - this should cover all frontend, command line and alerting data requests
    - [`data query type`](https://github.com/grafana/grafana-plugin-sdk-go/blob/74f14dae307677ddf08c7caca09575dc4d594df1/backend/data.go#L53) is available and useful
  - Cons:
    - no status code and duration is available since it's logged before response

For both alternatives, we may want to add a configuration toggle for enabling the logging in ini file to avoid verbose logging by default.

Comments ?

**Reply by marefr** [reply]
> so we are really requesting this feature from **logs** perspective.

I see. If Grafana would log these attributes in a structured manner in its server log would that be an option for you as well? 

> * printing logs in [`query` service](https://github.com/grafana/grafana/blob/a83fdc6b879127d506f0ffc3639a25cdf36bf246/pkg/services/query/query.go#L224) before sending out to individual data source

Ohh I see now that this is an option you suggest. If we would log this using our contextual logger it would include a `traceID` and that same `traceID` would also be included in the HTTP request log (that includes status code and duration) allowing you to correlate things. Currently requires to have distributed tracing configure for that `traceID` to show up/be included. 

Please note that the response status from /api/ds/query soon will be 207 Multi Status based on https://github.com/grafana/grafana/pull/48550 changes, but it's currently behind a feature toggle.

**Comment by ryantxu** [comment]
Check https://github.com/grafana/grafana/pull/48550 -- that has been merged

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). Due to the age and number of responses to this discussion, we are deciding to close it. If this is something you would like to see in Grafana, feel free to open an [issue](https://github.com/grafana/grafana/issues) so the discussion can continue. Thank you!
